### ISSUE TITLE
incorrect & changing results for transformation along modulo axis

### ISSUE THREAD

**Issue by AndrewWittenberg** [NOAA-PMEL/Ferret#1916]
Here's a scary bug identified by Sergey Malyshev at GFDL, while working with CMIP5 data.  I've boiled it down to a simpler example that needs no input datasets.

Save the following script as "mytest.jnl":

```
\can mode verify
! mytest.jnl
! Usage: go mytest [region]
!                  [  $1  ]
!
! Examples:
! yes? go mytest x=-10:30  ! example_1.png
! yes? go mytest x=10:50   ! example_2.png

def sym my_reg "$1%x=-10:30%"

def ax/x=0:358.75/np=288/unit="degrees_east" lon
let a = randn(x[gx=lon])
let a_tot = a[x=@sum]
let w  = a/a_tot

plot/vl=-4:4/($my_reg)/title="region: ($my_reg)" w/0
plot/ov/($my_reg)/line/col=1/sym=18/title="1" w

load/x=-90:30 w
plot/ov/($my_reg)/line/col=2/sym=2 w

load/x=-90:40 w
plot/ov/($my_reg)/line/col=3/sym=4 w

load/x=-359:359 w
plot/ov/($my_reg)/line/col=4/sym=28 w

set mode/last verify
```

Then run:

```
 	NOAA/PMEL TMAP
 	PyFerret v7.44 (optimized)
 	Linux 3.10.0-957.el7.x86_64 - 12/10/18
 	15-Mar-19 19:57     

yes? go mytest x=-10:30
yes? frame/file=example_1.png
```

![example_1](https://user-images.githubusercontent.com/34082818/54467852-2a149280-475e-11e9-8d40-1792629c9b96.png)

```
yes? go mytest x=10:50
yes? frame/file=example_2.png
```

![example_2](https://user-images.githubusercontent.com/34082818/54467857-37318180-475e-11e9-8d93-cce38cc4518f.png)

Here the LOADs should be doing nothing, since we specify the region explicitly in the PLOT command.  But somehow they're contaminating the context of each plot.  In each figure, all 4 curves should lie exactly on top of each other, but they don't!  (In Sergey's original example, the LOADs were effectively done by other plots.)

I also tried this with the old vanilla Ferret v6.82 (dated 8/3/2012), and saw similar behavior, though the curves looked a little different.  It's hard to believe that we haven't caught this before. 

This appears to produce incorrect answers without warning -- so please put this at the top of the "urgent & critical" stack.  We'd appreciate an updated executable ASAP.

Possible leads: issues #1914, #463, #1768, #1849.

**Comment by AnsleyManke** [comment]
Ok , wow.  If we do a SET MODE STUPID so the code never re-uses results in memory, then as you say the curves lie on top of each other.

We're on it!

**Comment by karlmsmith** [comment]
Another interesting discovery: 
if you call with X=10W:30E they all overlap,
OR
if the load command use W and E they all overlap

**Comment by karlmsmith** [comment]
Well, most work.  10E:50E still the same plot with the fourth off.

**Comment by karlmsmith** [comment]
Also defining the axis using /X=0:358.75E (and leave out the units) can make a difference in the plots produced.  But always one or the other plot has a problem.

Ansley mentioned diagnostics show differences in the indices when trying to resolve coordinates.

**Comment by karlmsmith** [comment]
The issue is what it means to evaluate over a region.  This can more easily be seen from listings instead of plots.  Ferret is doing things correctly; it is the syntax messing us up here.

When ferret can get the data for the region specified to show from cache, it grabs that data and uses it.  When you go outside the region to be shown, it has to re-evaluate for the specified region.  In this example, there are two places that change the expected results every time Ferret has to recalculate: `randn` and `@sum`.  

If Ferret has to re-evaluate `randn`, then is will pull up a fresh set of random values.  So switch this to something like `cos` to get consistent values every time the variable is evaluated.  There is still a mismatch, but this time it is due to what region you are evaluating `@sum`.  So for the last plot, you load the value from using the sum over the region -359 to 359.  For the plot, Ferret sees these values are already loaded and uses them.  But these are different values from using the sum over the smaller original region specified.

**Comment by AnsleyManke** [comment]
The thing to do to get a consistent result is redefine the variable so it's explicitly doing the sum over the desired region.  Changing the function in the example from randn to cos to get away from the oddities of the random, and putting the region over which to sum inside the definition of the variable a_tot gives us consistent results.


```
def ax/x=0:358.75/np=288/unit="degrees_east" lon
let a = cos(x[gx=lon])
let a_tot = a[($my_reg)@sum]
let w  = a/a_tot
```

Are there any documentation holes here?  Maybe an FAQ or note to the Users' List?  If this confused all of us, it's definitely one of those "thinking like a Ferret" issues that can trip us up.

**Comment by karlmsmith** [comment]
If we feel a programming solution is required, I can think of a couple of ways of prevent users getting caught in this mess.  I have not thought about how difficult each option would be to implement or its impact on existing scripts.

1. always force users to specify the region over which to do the transformation (ie, no default region)
2. set the default region to be the full length of the axis/axes regardless of region specified
3. do not cache variables that have a transformation (and variables using those variables)

**Comment by karlmsmith** [comment]
Another possibility:

4. setting a region, either explicitly with `SET REGION` or implicitly using `/X= /Y= ...` qualifiers, always clears the cache before evaluation

**Comment by AndrewWittenberg** [comment]
Thanks for looking into this.  Sorry that my efforts to simplify the report introduced unnecessary complications, due my using RANDN() to create an example field instead of just sending you a NetCDF file.  (I think that what confused Sergey and me is a mixture of the above, and possibly a different bug having to do with modulo longitudes.  See my revised example below.)

Ansley's recommendation to make the context explicit in the variable definition is well noted -- I'll relay that to Sergey.

Continuing the above thread from my original report... If I'm interpreting correctly, Ferret re-uses the cached result whenever the current context is a *subset* of a previously cached context. I was aware of this for simple un-transformed variables; e.g. re-plotting such a variable for a zoomed-in region.  But I had thought that if the variable involved transformations over default regions, Ferret would re-evaluate those transformations when the region changed.  In such a case, shouldn't "thinking like a Ferret" imply that the variable is always evaluated with respect to the latest context?  Otherwise (as in this example) we'd get Frankenstein results involving mixtures of current and past contexts.

Regarding implementation: I'd avoid Karl's 1-3, which would be fundamental changes from Ferret's philosophy.  But maybe something like 4: when evaluating a variable that involves region-dependent transformations (i.e. not smoothers), only reuse the cached result if the current region is *identical* to (not just a subset of) the region used to  compute the cached result.

And now... here's a report that's closer to what brought Sergey to my office.  You'll see why were were so confused.  Here's a gzipped "a.nc": [a.nc.gz](https://github.com/NOAA-PMEL/Ferret/files/2980056/a.nc.gz)

```
 	NOAA/PMEL TMAP
 	PyFerret v7.44 (optimized)
 	Linux 3.10.0-957.el7.x86_64 - 12/10/18
 	18-Mar-19 15:59     

yes? use "a.nc"
yes? let w = areacella/areacella[x=@sum,y=@sum]
yes? can view
yes? set view ul; shade/x=-10:30 w             !why the shift at x=0?
yes? set view ur; shade/x=-90:30 w             !now it shifts the other way!
yes? set view ll; shade/x=-10:30 w             !and this has changed the original plot!
yes? set view lr; shade/x=-359:359 w           !and now the shift disappears?!
yes? frame/file=example_3.png
```
![example_3](https://user-images.githubusercontent.com/34082818/54560247-48fe6900-4998-11e9-9bcf-d57b430e47a7.png)

**Comment by AnsleyManke** [comment]
Ferret does check the context related to transformations when it's deciding if it's got a previously cached reusable result.  We're getting there; thank you for the further example

**Comment by karlmsmith** [comment]
What you are seeing - the bug - is that it appears to be splitting the calculation at the modulo break, but doing the sum only over the portion in that part instead of doing the sum over the whole region.  So the list of `/X=-10:30` is the list of `/X=-10:-1.25` appended by the list of `/X=0:30` as can be seen by:

```
let w = 1.0E5 * areacella/areacella[x=@sum,y=@sum]
list /width=500 /prec=7 /x=-10:30 w             !why the shift at x=0?
can mem /all
list /width=500 /prec=7 /x=-10:-1.25 w
can mem /all
list /width=500 /prec=7 /x=0:30 w
```

The last plot appears to be correct but is not; the regions are approximately the same size but there is a numeric change if listed.

Ansley has found this error starts appearing about the time vtree was introduced, as is digging into the changes made.

**Comment by karlmsmith** [comment]
Note that the following works fine:
```
! insignificantly different from a constant, but make sure no short-cuts
let factor = 1.0E14 + 100.0 * _X[gx=areacella] + 100.0 * _Y[gy=areacella]
let w = factor / areacella[x=@sum,y=@sum]

list /width=500 /prec=7 /x=-10:30 w
can mem
list /width=500 /prec=7 /x=-10:-1.25 w
can mem
list /width=500 /prec=7 /x=0:30 w
```
(the first giving all 1.710895, the second all 7.05744, the third all 2.258382).
So something about using `areacella` in the definition of `w` changes the behavior.

**Comment by karlmsmith** [comment]
And a note on the caching.   (This is probably a topic for a new ticket.)

If the `can mem` are left out, the cached values from the first listing are re-used (so everything is 1.710895).  But this brings up the question of what does the `/X=...` option really mean?

If it is indeed an on-the-fly definition of a new region, then anytime these (as well as explicit region specifications) are made, then any cached user-defined variables need to be cleared.

Or is it "give me just these values for the current region definition", in which case the cache results should (must) be used.

If the former, how does one accomplish the latter?

Of course, all of this arises as a result of not specifying the range for `@SUM`.  If `[X=0:360@SUM,Y=-90:90@SUM]` is used, all this ambiguity goes away (and every value become the same in the example immediately above).  **And the original problem goes away as well.**

**Comment by karlmsmith** [comment]
Forgot to mention that using `set region /x=... /y=...; list w` and `list w[X=...,Y=...]` in the example immediately above gives the the same results.  My testing script: [tests.jnl.txt](https://github.com/NOAA-PMEL/Ferret/files/2983025/tests.jnl.txt)

**Comment by AndrewWittenberg** [comment]
Regarding what `/x=...` means: in the documentation on "thinking like a Ferret"

https://ferret.pmel.noaa.gov/Ferret/documentation/users-guide/introduction/GETTING-STARTED#_VPID_7

under "Contexts," it says: "Command context information on any axis or on the data set will replace any program context information on the same axis or the data set."

So I'd interpret this broadly, as an on-the-fly application of a new region.  We wouldn't need to clear *all* cached variables, just the ones required by the evaluation that involve both default contexts and region-dependent transformations (like `@ave`, `@sum`, etc.).

If we wanted to just "zoom into values for the current region," then we'd need to supply an explicit region in the variable definition, e.g. `[x=0:360@sum,y=-90:90@sum]`.  Then we could zoom into it unchanged.

(This is actually how I thought Ferret worked all along -- hence my surprise about how the caching worked.)

**Comment by AnsleyManke** [comment]
Ferret does do all of this right.  Ferret executables prior to v6.85  were correct. I have tracked this bug down to the change made for #1323. 

That change had the effect of changing the order of operations for handling the modulo-ing, averaging, putting the expression all back together.

I think the answer here is to restore the prior behavior and reopen #1323. The `RETURN=`  code attempts to find things out about expressions without fully loading them. It needs a different fix.

**Comment by AnsleyManke** [comment]
This is fixed and checked in, routine merge_grid.F

**Comment by AndrewWittenberg** [comment]
Thanks so much for tracking this down.  We can wait on the fix to #1916 until #1323 is addressed, since we've made a lot of use of the new behavior as fixed by #1323.

**Comment by karlmsmith** [comment]
The latest fix for #1323 still allows this problem to reappear if
```
say `w,ret=shape`
```
is put in front of all the evaluations; i.e., going down the path for #1323 leaves the incorrect context around so the evaluation ends up doing what it was doing before.

**Comment by AnsleyManke** [comment]
aha. Good catch.  I'll check into that.

**Comment by AnsleyManke** [comment]
My fix earlier last week was fine as far as it went but missed a couple of things.  I now have a complete fix - pending gfdl testing

-for our future reference - 
When we do  RETURN=*size or shape,  the grids are analyzed but data isn't loaded.  The context information that's created and stored in cx_* data structures while doing this is not re-used. 

During evaluation of grids in RETURN=,  the grids associated with variables and expressions are stored, in a generic kind of way. For instance all we need is that a SUM  or other compressing transformation takes the size for that part of the expression in that direction down to size of 1, so there are things that are handled as if data is on the full-size grids of all underlying variables. The grid found is stored by IS_UVAR_GRID, making calls to SAVE_UVAR_GRID. This call, storing grid info in the linked-list dataset structure, was a change to allow for larger numbers of grids, along with dynamic memory management.  It wasn't taken into account correctly in the previous work on for this bug and #1323, possibly also #1545, and so grid info computed in this partial way was held onto and applied later when expressions are actually being computed.

When cleaning up at the end of repl_exprns, need to clear that stored grid information and do general user-variable leanup in a call to DEALLO_DYN_GRIDS . The grid of the expression will be determined afresh if the expression is evaluated and loaded into memory with whatever context is in place at that time. 

Finally, a CANCEL MEMORY also needed to go and clear out user-variable information that had been set in save_uvar_grid.

( memory-resident data and its' context is not touched by these changes.)

**Comment by AnsleyManke** [comment]
still one issue when working with Discrete Sampling Geometries files.

**Comment by AnsleyManke** [comment]
This is fixed. No changes needed in the main Ferret/PyFerret code
